src/dotnet/APIView/ClientSPA/src/app/_components/api-revision-options/api-revision-options.component.ts

```diff
@@ -191,74 +194,89 @@ export class ApiRevisionOptionsComponent implements OnChanges {
-    const semVarRegex = /(?<major>0|[1-9]\d*)\.(?<minor>0|[1-9]\d*)\.(?<patch>0|[1-9]\d*)(?:(?<presep>-?)(?<prelabel>[a-zA-Z]+)(?:(?<prenumsep>\.?)(?<prenumber>[0-9]{1,8})(?:(?<buildnumsep>\.?)(?<buildnumber>\d{1,3}))?)?)?/;
-
-    let latestGAApiRevision : any = null;
-    let latestApprovedApiRevision : any = null;
-    let currentMainApiRevision : any = null;
-    let latestReleasedApiRevision : any = null;
-
-    while (mappedApiRevisions.length > 0) {
-      let apiRevision = mappedApiRevisions.shift();
+  tagSpecialRevisions(mappedApiRevisions: any []) {
+    this.tagLatestGARevision(mappedApiRevisions);
+    this.tagLatestApprovedRevision(mappedApiRevisions);
+    this.tagCurrentMainRevision(mappedApiRevisions);
+    this.tagLatestReleasedRevision(mappedApiRevisions);   
+  }
 
-      if (latestGAApiRevision === null) {
-        if (apiRevision.version) {
-          let versionParts = apiRevision.version.match(semVarRegex);
+  tagLatestGARevision(apiRevisions: any[]) {
+    const gaRevisions : any [] = [];
 
-          if (versionParts.groups?.prelabel === undefined && versionParts.groups?.prenumber === undefined &&
-            versionParts.groups?.prenumsep === undefined && versionParts.groups?.presep === undefined) {
-              apiRevision.isLatestGA = true;
-              latestGAApiRevision = apiRevision;
-              continue;
-          }
+    for (let apiRevision of apiRevisions) {
+      if (apiRevision.version) {
+        const semVar = new AzureEngSemanticVersion(apiRevision.version, apiRevision.language);
+        if (semVar.versionType == "GA") {
+          apiRevision.semanticVersion = semVar;
+          gaRevisions.push(apiRevision);
         }
       }
+    }
 
-      if (latestApprovedApiRevision === null) {
-        if (apiRevision.isApproved) {
-          apiRevision.isLatestApproved = true;
-          latestApprovedApiRevision = apiRevision;
-          continue;
-        }
-      }
+    if (gaRevisions.length > 0) {
+      gaRevisions.sort((a: any, b: any) => a.semanticVersion.compareTo(b.semanticVersion));
+      gaRevisions[0].isLatestGA = true;
+      this.updateTaggedAPIRevisions(apiRevisions, gaRevisions[0]);
+      return gaRevisions[0];
+    }
+  }
 
-      if (currentMainApiRevision === null) {
-        if (apiRevision.apiRevisionType === 'Automatic') {
-          apiRevision.isLatestMain = true;
-          currentMainApiRevision = apiRevision;
-          continue;
-        }
-      }
+  tagLatestApprovedRevision(apiRevisions: any[]) {
+    const approvedRevisions : any [] = [];
 
-      if (latestReleasedApiRevision === null) {
-        if (apiRevision.isReleased) {
-          apiRevision.isLatestReleased = true;
-          latestReleasedApiRevision = apiRevision;
-          continue;
-        }
+    for (let apiRevision of apiRevisions) {
+      if (apiRevision.isApproved) {
+        apiRevision.approvedOn = apiRevision.changeHistory.find((change: any) => change.changeAction === 'approved').changedOn;
```

Comment:
Can you please make sure change history for approved status is present? This will throw a null reference error if an approved revision is missing approval change history (possible for historical data if approval history was not correctly rebuilt). I understand that most of the records will have this property but it's safer to handle this condition.

Reply:
Updated.